R/Analysis.R

```diff
@@ -443,5 +443,127 @@ CalculateClusterEnrichment <- function(seuratObj,
+#' 
+#' @examples
+#' \dontrun{
+#' #set the seurat Idents for FindAllMarkers
+#' Seurat::Idents(seuratObj) <- "ClusterNames_0.2"
+#' 
+#' markers <- Seurat::FindAllMarkers(seuratObj)
+#' 
+#' #filter markers to display the largest cluster identity markers according to average log fold change and differences in pct expression.
+#' strong_markers <- markers[abs(markers$avg_log2FC) > 3 & abs(markers$pct.1 - markers$pct.2) > 0.25, "gene"]
+#' 
+#' dotPlot <- ClusteredDotPlot(seuratObj, features = strong_markers, groupFields = "ClusterNames_0.2", scaling = 'row', ggplotify = TRUE)
+#' print(dotPlot)
+#' 
+#' }
+
+
+ClusteredDotPlot <- function(seuratObj, features, groupFields = "ClusterNames_0.2", assay = "RNA", ggplotify = TRUE, scaling = 'row', layer = 'data') {
+  #Sanity checks
+  #If you do some filtering upstream that removes all of the genes in your features vector, this doesn't error in an obvious way, so throw a specific error if you feed an empty vector into the features argument.
+  if (length(features) == 0) {
+    stop("The features argument is empty. Please specify a non-empty vector of features.")
+  }
+  #check that features are both valid and force them to be unique. 
+  if (!all(features %in% rownames(Seurat::GetAssayData(seuratObj, slot = 'data')))) {
+    warning(paste0('Features not found in Seurat object: ', paste0(features[!features %in% rownames(Seurat::GetAssayData(seuratObj, layer = layer))], collapse = ', ')))
+  }
+  if (any(duplicated(features))) {
+    warning(paste0('Features supplied are not unique. Duplicates will be removed.'))
+    features <- unique(features)
+  }
+  #check that groupFields are in the metadata
+  if (!all(groupFields %in% colnames(seuratObj@meta.data))) { 
+    stop(paste0('The following groupFields were not found in Seurat object metadata: ', paste0(groupFields[!groupFields %in% colnames(seuratObj@meta.data)], collapse = ', ')))
+  }
+  #check that scaling is supported
+  if (!scaling %in% c('row', 'column', 'none')) {
+    stop(paste0('Scaling method not supported: ', scaling, '. Please use one of: "row", "column", or "none".'))
+  }
+  #check assay
+  if (!assay %in% Seurat::Assays(seuratObj)) {
+    stop(paste0('Assay not found in Seurat object: ', assay))
+  }
+  #check ggplotify boolean
+  if (!is.logical(ggplotify)) {
+    stop(paste0('ggplotify: ', ggplotify, ' is not a boolean. Please specify ggplotify = TRUE or ggplotify = FALSE. If TRUE, the ComplexHeatmap object will be converted to a ggplot object.'))
+  }
+  
+  #create averaged Seurat object for mean expression
+  avgSeurat <- Seurat::AverageExpression(seuratObj, 
+                                         group.by = c(groupFields),
+                                         layer = layer, 
+                                         return.seurat = T,
+                                         assays = assay)
+
+  #scale, if requested
+  if (scaling != "none") {
+    mat <- as.matrix(Seurat::GetAssayData(avgSeurat, layer = layer))  %>%
```

Comment:
this is really picky, but shouldnt you subset before you scale? practically it probably doesnt matter that much, but it's wasted computation

Reply:
true - fixed.